File: Java/92. Reverse Linked List II.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        if (left == right || head == null)  return head;
        
        ListNode prev = null;
        ListNode cur = head, next;
        
        // reach to the left position
        while (left > 1) {
            left--;
            right--;
            prev = cur;
            cur = cur.next;
        }
        
        ListNode l = cur, p = prev;
        while (right > 0) {
            right--;
            next = cur.next;
            cur.next = prev;
            prev = cur;
            cur = next;
        }
        
        if (p != null) {
            p.next = prev;
        } else {
            head = prev;
        }
        
        l.next = cur;
        return head;
    }
}

File: Python3/92. Reverse Linked List II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if left == right:
            return head
        
        prev = ListNode()
        prev.next = head
        ph = prev
        cur = head
        count = 1
        
        while count < left:
            prev, cur = cur, cur.next
            count += 1
            
        tail = cur
        prev.next, end = self.reverse(cur, right-count)
        tail.next = end
        return ph.next
        
    def reverse(self, left: ListNode, right: int):
        prev = None
        cur = left
        count = 0
        while left and count < right:
            prev, left.next, left = left, prev, left.next
            count += 1
            
        # now left points to right value's node
        nxt = left.next
        left.next = prev
        return left, nxt